connectors/slack/template/index.js

```diff
@@ -1,24 +1,295 @@
+    });
+    return;
+  }
+
+  storage.ensureStorage(ctx);
+
+  storage.get().then(() => {
+    const [bearerString, token] = ctx.headers.authorization.split(' ');
+    if (bearerString !== 'Bearer' || !Object.keys(storage.tokenToTeam).includes(token)) {
+      cb(null, {
+        status: 401,
+        body: { message: 'Invalid auth token' },
+      });
+      return;
+    }
+
+    let team, handler;
+    if (!isDeregister) {
+      if (!ctx.body || !ctx.body.app || !ctx.body.team_id || !ctx.body.handler) {
+        cb(null, {
+          status: 401,
+          body: { message: 'Missing required app, team_id, and handler body parameters' },
+        });
+        return;
+      } else {
+        team = `${ctx.body.app}-${ctx.body.team_id}`;
+        handler = ctx.body.handler;
+      }
+    } else {
+      if (!ctx.query || !ctx.query.app || !ctx.query.team_id || !ctx.query.handler) {
+        cb(null, {
+          status: 401,
+          body: { message: 'Missing required app, team_id, and handler query parameters' },
+        });
+        return;
+      } else {
+        team = `${ctx.query.app}-${ctx.query.team_id}`;
+        handler = decodeURIComponent(ctx.query.handler);
+      }
+    }
+
+    const tokenUsedUp = 'TokenUsedUp';
+    const registrationProblem = 'RegistrationProblem';
+
+    storage
+      .put(() => {
+        // Doing the token use check in here in case there is a storage conflict
+        // and when the data is refreshed, it turns out the token was used up
+
+        const [storedTeam, remainingUses] = storage.tokenToTeam[token];
+        if ((!isDeregister && storedTeam !== team) || remainingUses < 1) {
+          Sdk.debug('Token is used up');
+          let e = new Error();
+          e.name = tokenUsedUp;
+          throw e;
+        }
+
+        let handlerRegistered = storage.teamToHandler[team] && storage.teamToHandler[team].includes(handler);
+        if ((!isDeregister && handlerRegistered) || (isDeregister && !handlerRegistered)) {
+          let e = new Error();
+          e.name = registrationProblem;
+          throw e;
+        }
+
+        if (!isDeregister) {
+          Sdk.debug('Registering', handler, team);
+          if (!storage.teamToHandler[team]) {
+            storage.teamToHandler[team] = [];
+          }
+          storage.teamToHandler[team].push(handler);
+        } else {
+          Sdk.debug('Deregistering', handler, team);
+          storage.teamToHandler[team].pop(handler);
+          if (storage.teamToHandler[team].length == 0) {
+            delete storage.teamToHandler[team];
+          }
+        }
+
+        if (remainingUses == 1) {
+          Sdk.debug('Token last use');
+          delete storage.tokenToTeam[token];
+        } else {
+          Sdk.debug('Token first use');
+          storage.tokenToTeam[token] = [team, remainingUses - 1];
+        }
+      })
+      .then(() => {
+        cb(null, {
+          status: isDeregister ? 200 : 201,
```

Comment:
Why 201? I think it is always 200, the registration is fully completed when you return.

Reply:
Fixed